apollo-router/src/uplink/entitlement_stream.rs

```diff
@@ -135,34 +136,48 @@ where
                         // We got a new claim, so clear the previous checks.
                         this.checks.clear();
-                        // Insert the new checks.
-                        this.checks.insert_at(
-                            Event::WarnEntitlement,
-                            (to_positive_instant(claim.warn_at)).into(),
-                        );
-                        this.checks.insert_at(
-                            Event::HaltEntitlement,
-                            (to_positive_instant(claim.halt_at)).into(),
-                        );
-                        next
+                        let claims = entitlement.claims.as_ref().expect("claims is gated, qed");
+                        let halt_at = to_positive_instant(claims.halt_at);
+                        let warn_at = to_positive_instant(claims.warn_at);
+                        let now = Instant::now();
+                        // Insert the new checks. If any of the boundaries are in the past then just return the immediate result
+                        if halt_at > now {
+                            // Only add halt if it isn't immediately going to be triggered.
+                            this.checks.insert_at(
+                                Event::UpdateEntitlement(EntitlementState::EntitledHalt),
+                                (halt_at).into(),
+                            );
+                        } else {
+                            return Poll::Ready(Some(Event::UpdateEntitlement(
+                                EntitlementState::EntitledHalt,
+                            )));
+                        }
+                        if warn_at > now {
+                            // Only add warn if it isn't immediately going to be triggered and halt is not already set.
+                            // Something that is halted is by definition also warn.
+                            this.checks.insert_at(
+                                Event::UpdateEntitlement(EntitlementState::EntitledWarn),
+                                (warn_at).into(),
+                            );
+                        } else {
+                            return Poll::Ready(Some(Event::UpdateEntitlement(
+                                EntitlementState::EntitledWarn,
+                            )));
+                        }
+
+                        Poll::Ready(Some(Event::UpdateEntitlement(EntitlementState::Entitled)))
                     }
                     // Upstream has a new event with no claim
                     (Poll::Ready(Some(_)), _) => {
                         // As we have no claim clear the checks
                         this.checks.clear();
-                        next
+                        Poll::Ready(Some(Event::UpdateEntitlement(EntitlementState::Unentitled)))
                     }
                     // There were still active checks, so go for pending
                     (Poll::Ready(None), true) => Poll::Pending,
```

Comment:
if we return `Pending` but the underlying stream has not been polled a second time, then the entitlement stream will never be polled again.

the pattern here should be something like:

```rust
loop {
  match stream.poll_next() {
    Poll::Pending => return Poll::Pending,
    Poll::Ready(pattern) => return Poll::Ready(something),
    // this will go through the loop again to call poll_next again and register interest in the underlying stream
    Poll::Ready(other pattern) => {}
  }
}
```

Reply:
that's the same on line 138 in the check for poll_expired, there's a risk it will just stop polling

Reply:
I think I have a decent solution in: [0f82260](https://github.com/apollographql/router/pull/2596/commits/0f8226041974176ae7e513637af2b008505ad336)

active_checks has been removed so that it is easier to see how we are handling things.
In addition, every `Poll::Pending` has a comment to indicate where the waiter will have been scheduled.

Reply:
Also FuseStream is used to prevent calling of upstream if it has been terminated.

Reply:
@garypen Could use your eyes on this as well.

Reply:
Significant refactoring occurred and we had a sync walkthrough to resolve this.